# TD1 Python - Premiers pas avec python

Vous trouverez par la suite des exercices obligatoires et des exercices pour approfondir. Les exercices obligatoires sont
précédés par le caractère *.

[Qu'est-ce que notebook ?](http://nbviewer.jupyter.org/github/ipython/ipython/blob/3.x/examples/Notebook/Index.ipynb)

# Identifiez vous 

In [5]:
print("Creation d'une session de suivi sous https://courdier.pythonanywhere.com/records...")
import requests
exec(requests.get("https://courdier.pythonanywhere.com/get-send-code").content)
npt_config = {
    'session_name': 'python_td1',
    'session_owner': 'courdier',
    'sender_name': input("Your name:")',
}
send("starting TD1", 1)

Creation d'une session de suivi sous https://courdier.pythonanywhere.com/records...


<Response [200]>

In [2]:
from answers_td1 import *
init_exo_1()

Creation d'une session de suivi sous https://courdier.pythonanywhere.com/records...


/Users/remycourdier/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:34: NotOpenSSLWarning: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


NameError: name 'send' is not defined

# Installation de l'environnement de travail Python

## sur Windows et OSX
Télécharger et suivez la procédure d'installation de la dernière version de python 3 sur le site https://www.python.org. 
Vous trouverez aussi des indications pour l'installation sur ce site http://docs.python-guide.org/en/latest/

## Sur Linux : 
Si vous voulez avoir l'environnement correspondant à celui spécifié dans le TD1, l'installation de python peux se faire avec les commandes suivantes (dans une fenêtre shell) :
<pre>
sudo apt-get install python3 
sudo apt-get install idle3
</pre>

par ailleurs, sur votre distribution de Linux dans "applications / éducation / mathématiques" vous devriez avoir un IDE appelé "Eric Python IDE".
Il est possible de travailler également avec cet outil pour les TD de ce cours.

Vous pouvez passer également par le market de Ubuntu : "Logiciels Ubuntu". Puis rechercher "python" dans la barre de recherche.

# Exercice 1 : Utilisez l'interpreteur

## 1.1 : Faire des calculs

### *Exercice 1.1

1) Evaluer les expressions suivantes et obervez les résultats calculés par l'interpréteur python :

In [2]:
5+3

8

In [3]:
2 - 9

-7

In [4]:
5 + 3 * 2

11

In [5]:
(5+3) * 2

16

In [6]:
20 / 3

6.666666666666667

In [7]:
20 // 3

6

In [8]:
20.0 // 3

6.0

In [10]:
25 % 3

1

In [11]:
sqrt (2)

NameError: name 'sqrt' is not defined

L'interpréteur Python exécute l'instruction lorsque vous frappez sur la touche entrée ( ). Faites bien attention à la valeur retournée par l'interpréteur.
* Python respecte-t-il les prioritées des opérations arithmétiques ?

In [ ]:
rep111

* Que se passe-t-il lorsqu'on mélange des entiers et des flottants dans un même calcul ?

In [ ]:
print(rep112)

* Que font les operateurs / et // ?

In [ ]:
print(rep113)

* Que calcule l'opérateur % ?

In [ ]:
print(rep114)

### *Exercice 1.2

1) Tapez et exécutez les expressions suivantes et observez et analyser la réponse de l'interpréteur.

In [ ]:
pi

In [12]:
from math import *

In [13]:
sqrt (2)

1.4142135623730951

In [14]:
sqrt 2

SyntaxError: invalid syntax (<ipython-input-14-0c9b627e8603>, line 1)

In [15]:
sin (90)

0.8939966636005579

In [16]:
pi

3.141592653589793

In [17]:
sin(pi /2)

1.0

In [18]:
sin (radians (90))

1.0

Pour connaitre les fonctions qui sont définies dans un module, il faut utiliser la commande help("nom_du_module").
L'interpréteur affiche alors le descriptif de toutes les fonctions qu'il contient (dans un terminal tapez la barre d'espace
pour afficher la suite et q pour quitter l'affichage). Pour obtenir l'aide sur une fonction en particulier, tapez (après
avoir importée le module dans l'environnement) help(fonction) sans les guillemets !

2) Essayez les commandes help("math") et help(sqrt) (apres avoir execute from math import *).

In [ ]:
help("math")

In [ ]:
help (sqrt)

### Exercice 1.3

1) Evaluez les expressions suivantes dans l'interpréteur :

In [ ]:
13 // 3

In [24]:
14 % 3 # 3 x 4 = 12 ; 14-12 = 2 ; reste de la division entière

2

In [22]:
10 ** 2

1000

In [ ]:
2 ** 0.5

2) Comment savoir si un nombre est impair ?

In [ ]:
print(rep132)

## 1.2 Utiliser des variables

### *Exercice 1.4

Décrivez le plus complètement possible ce qui se passe lors de l'évaluation de chacune des instructions suivantes :

In [25]:
n = 7
rac2 = 1.414
x = n + rac2
print (n + rac2)
print (x)

8.414
8.414


In [26]:
msg = " Rien a signaler "

In [27]:
msg

' Rien a signaler '

In [28]:
msg2 = msg

In [29]:
print ( msg2 )

 Rien a signaler 


In [30]:
msg3 = print ( msg2 )

 Rien a signaler 


In [31]:
msg3

In [32]:
print ( msg3 )

None


### *Exercice 1.5

Testez les expressions suivantes :

In [35]:
a=10 ; b=33
a < b

True

In [36]:
b < a

False

In [37]:
a+23 = b

SyntaxError: can't assign to operator (<ipython-input-37-6d57f7bf1ce1>, line 1)

In [38]:
a+23 == b

True

In [39]:
c = b == a + 23

In [40]:
c

True

In [41]:
" abcd " < "bbc"

True

In [42]:
"d" < ""

False

In [45]:
"33" > "323"

False

1) Comment s'effectue la comparaison de chaines de caractères ?

In [ ]:
print(rep151)

### Exercice 1.6

1) Testez les instructions suivantes :

In [46]:
int (12)

12

In [48]:
a=12
float (a)

12.0

In [49]:
int (2.71)

2

In [50]:
c=float (" 10.7 ")
c

10.7

In [52]:
int(" 10.7 ")

10

In [68]:
a="10"
b=5
int(a) + int(b)

15

In [69]:
str(a) + str(b)

'105'

2) Que font les fonctions int et float ?

In [ ]:
print(rep162)

### Exercice 1.7

Pour chacune des expressions booléennes suivantes affectez les valeurs **True** ou **False** aux variables a, b, c de telle sorte que la valeur de l'expression soit vérifiée (**True**).

In [71]:
print ("a or b and c")
a=(input("valeur de a : ") == "True") and True or False 
b=(input("valeur de b : ") == "True") and True or False 
c=(input("valeur de c : ") == "True") and True or False
print("==> " + str(a) + " or " + str(b) + " and " + str(c))
a or b and c

a or b and c
==> True or False and False


True

In [72]:
print ("a and not b or c")
a=(input("valeur de a : ") == "True") and True or False
b=(input("valeur de b : ") == "True") and True or False 
c=(input("valeur de c : ") == "True") and True or False
print("==> " + str(a) + " and not " + str(b) + " or " + str(c))
if a and not b or c :
    print("condition verifiée")
else : 
    print("condition non verifiée")

a and not b or c
==> True and not False or True
condition verifiée


In [73]:
print ("a and not(b or a)")
a=(input("valeur de a : ") == "True") and True or False
b=(input("valeur de b : ") == "True") and True or False 
a and not(b or a)

a and not(b or a)


False

In [74]:
print ("(a or b) and not(a and b)")
a=(input("valeur de a : ") == "True") and True or False
b=(input("valeur de b : ") == "True") and True or False 
(a or b) and not(a and b)

(a or b) and not(a and b)


True

### Exercice 1.8

1) Testez les instructions suivantes

In [ ]:
r = 5
pi = 3.1415927
s = pi * r ** 2
type (r)

In [ ]:
type (pi)

In [ ]:
type (s)

In [ ]:
type ("1234")

In [ ]:
type (sqrt)

2) A quoi sert la fonction type ?

In [ ]:
print (rep182)

## 1.3 Instructions conditionnelles

### *Exercice 1.9

1) Tapez les instructions suivantes dans l'interpréteur de Python en prenant bien soin de taper une tabulation après les 3 points affchés par l'interpréteur lorsque c'est nécessaire. Les tabulations sont indispensables en langage Python et délimitent les blocs d'instructions !

In [79]:
a = 20
if a > 100 :
    # ce bloc est exécuté si l'expression (a > 100) est vraie
    print ("a depasse la centaine ")

2) Recommencez la même instruction conditionnelle mais en affectant la valeur 20 à la variable a. Que se passe-t-il ?

In [ ]:
print (rep192)

In [80]:
a = 20
if a > 100 :
    # ce bloc est exécuté si l'expression (a > 100) est vraie
    print ("a depasse la centaine ")
else :
    # ce bloc est exécuté si l'expression (a > 100) est fausse
    print("")
    print ("a ne depasse pas la centaine ")


a ne depasse pas la centaine 


### exercice 1.10

Tapez une séquence d'instructions pour affecter des valeurs (de votre choix) à deux variables x et y, puis à l'aide d'expressions d'alternative, afficher x est le plus grand ou bien y est le plus grand ou bien x égale y selon les valeurs de x et y. Faites bien attention aux tabulations pour indenter les blocs d'instructions. En particulier sivvous voulez passer des lignes à l'intérieur d'un bloc d'instructions vous devez quand m^eme tapez les tabulations sur les lignes vident.

In [81]:
x = 12
y = 23

if x > y :
    print ("x est le plus grand ")
else :
    if y > x :
        print ("y est le plus grand ")
    else :
        print ("x egale y")

y est le plus grand 


### Exercice 1.11

Tapez une séquence d'instructions pour affecter des valeurs à trois variables de votre choix puis, à l'aide d'alternatives, faites afficher le nom de la variable dont la valeur est la plus grande.

In [83]:
a=10 ; b=30 ; c=20
if a > b :
    if a > c :
        print ("a est le plus grand ")
    else :
        print ("c est le plus grand ")
else :
    if b > c :
        print ("b est le plus grand ")
    else :
        print ("c est le plus grand ")

c est le plus grand 


# 2. Premiers programmes

### *Exercice 1.12

Saisissez le programme suivant dans un éditeur de texte, enregistrez-le dans un fichier que vous nommez ex12.py (on utilise l'extension .py pour les programmes écrits en Python). Ne pas saisir les numéros de ligne.

In [102]:
# ceci est un commentaire
# programme exo12.py
# mon premier programme python

#  la commande input retourne un chaine de caractères qui doit etre convertie en entier
from random import * 
a = randint(-100,100)
if a > 0 :
    print("python a choisi la valeur aleatoire ", a)
else :
    a = int( input (str(a) + " inferieur à 0, Entrez un entier : "))
# On utilise des instructions conditionnelles imbriquées
if a < 0 :
    print ("Cet entier est negatif ") # affiche un resultat
else :
    if a > 0 :
        print ("Cet entier est positif ")
        if a < 100 :
            print ("et plus petit que 100")
        else :
            print ("et superieur ou egal a 100")
    else :
        print (" Vous avez entré 0")

python a choisi la valeur aleatoire  47
Cet entier est positif 
et plus petit que 100


### *Exercice 1.13

Ecrivez un programme qui affiche le menu suivant :
<pre>
1. Enregistrer la partie
2. Charger une partie
3. Nouvelle partie
4. Quitter

</pre>
puis demande à l'utilisateur de saisir son choix (l'utilisateur entre un entier) puis affiche OK si le choix est valide ou bien Choix incorrect sinon.

In [ ]:
# exemple d'exécution de cet exercice
test_exo_1_13()

In [103]:
# Voir le code source de cet exercice 1.13
test_exo_1_13??

Signature: test_exo_1_13()
Docstring: <no docstring>
Source:   
def test_exo_1_13():
    print("1. Enregistrer la partie")
    print("2. Charger une partie")
    print("3. Nouvelle partie")
    print("4. Quitter")
    ichoix = int(input("Votre choix : "))
    if (0 < ichoix and ichoix < 5): 
        print ("Ok")
    else:
        print("Choix incorrect")
File:      ~/Documents/GitHub/TP/TD1/answers_td1.py
Type:      function


### *Exercice 1.14

Ecrivez un programme qui demande la saisie de trois entiers et affiche **combien** de ces entiers sont égaux à zéro. Pensez à utiliser des opérateurs logiques dans vos conditions.

In [ ]:
# Exemple d'exécution de cet exercice 1.14 
test_exo_1_14()

In [104]:
# Voir le code source de cet exercice 1.14
test_exo_1_14??

Signature: test_exo_1_14()
Docstring: <no docstring>
Source:   
def test_exo_1_14():
    a = int( input (" Donner un premier entier "))
    b = int( input (" Donner un second entier ")) 
    c = int( input (" Donner un troisieme entier "))
    nb_zero = 0
    if a == 0 :
        nb_zero += 1
    if b == 0 :
        nb_zero += 1
    if c == 0 :
        nb_zero += 1
    print ("Il y a", nb_zero , " zeros ")
File:      ~/Documents/GitHub/TP/TD1/answers_td1.py
Type:      function


### *Exercice 1.15

Ecrivez un programme permettant de convertir une température donnée dans une unité (Celcius, Fahrenheit ou Kelvin)
dans les autres unités.
Les unités à considérer sont Celsius ( C ), Fahrenheit ( F ) et Kelvin ( K ) et les formules de conversion sont :
>**C** = 9/5(F-32)   
>**F** = 5/9C +32    
>**K** = C+273,15

Le programme demandera la valeur de la température puis l'unité dans laquelle elle est exprimée donnée sous la forme d'un caractère : C, F ou K.

In [105]:
# Exemple d'exécution de cet exercice : 
test_exo_1_15()

45.0 C = 57.0 F = 318.15 K


In [106]:
# Voir le code source de cet exercice 1.15
test_exo_1_15??

Signature: test_exo_1_15()
Docstring: <no docstring>
Source:   
def test_exo_1_15():
    # Programme de conversion de temperature
    # On saisit la valeur de la ytempetrature sous la forme d'un réel (float)
    temp = float ( input (" Donnez la temperature :"))
    unite = input (" Donnez l'unite C, F ou K : ")

    if ( unite != "C") and ( unite != "F") and ( unite != "K") :
        print ("Erreur , unite incorrecte ")
    else :
        if unite == "C" :
            tempC = temp
        elif unite == "F" :
            tempC = 9 * ( temp - 32) / 5.0
        elif unite == "K" :
            tempC = temp - 273.15
        
    tempF = 5 * tempC / 9.0 + 32
    tempK = tempC + 273.15
    print (tempC ,"C =", tempF , "F =", tempK , "K")
File:      ~/Documents/GitHub/TP/TD1/answers_td1.py
Type:      function


### Exercice 1.16

Ecrivez un programme qui demande la saisie, sous forme de flottant, de la longueur, de la largeur et de la hauteur
d'un parallélépipède rectangle et affiche son volume.

In [ ]:
# Exemple d'exécution de cet exercice : 
test_exo_1_16()

In [ ]:
# Voir le code source de cet exercice
test_exo_1_16??

### Exercice 1.17

Ecrivez un programme qui demande la saisie d'un entier et qui produit l'affichage Entier pair ou Entier impair selon que l'entier saisi est pair ou impair.

In [ ]:
# Exemple d'exécution de cet exercice : 
test_exo_1_17()

In [ ]:
# Voir le code source de cet exercice 1.15
test_exo_1_17??

### Exercice 1.18

Ecrivez un programme qui, selon le choix de l'utilisateur, permet de calculer l'aire et le périmètre d'un carré, d'un rectangle, d'un triangle rectangle (dont on connaît les longueurs des c^otés de l'angle droit ), d'un parallélogramme ou d'un losange.

In [31]:
# Exemple d'exécution de cet exercice : 
test_exo_1_18()

1. Carre 
2. Rectangle 
3. Triangle rectangle 
 Aire = 18.0
 Perimetre = 18.0


In [32]:
# Voir le code source de cet exercice
test_exo_1_18??

Signature: test_exo_1_18()
Docstring: <no docstring>
Source:   
def test_exo_1_18():
    # Affichage du menu
    print ("1. Carre ")
    print ("2. Rectangle ")
    print ("3. Triangle rectangle ")
    
    # Saisie du choix utilisateur
    figure = int( input (" Quelle figure :"))

    if figure == 1: # Carre
        cote = float ( input (" Donner la valeur du cote :"))
        print (" Aire = ", cote * cote )
        print (" Perimetre =", cote *4)
    elif figure == 2: # Rectangle
        longueur = float ( input (" Donner la valeur de la longueur :"))
        largeur = float ( input (" Donner la valeur de la largeur :"))
        print (" Aire =", longueur * largeur )
        print (" Perimetre =", 2*( longueur + largeur ))
    elif figure == 3: # Triangle
        cote1 = float ( input (" Donner la valeur du premier cote :"))
        cote2 = float ( input (" Donner la valeur du deuxieme cote :"))
        print (" Aire =", ( cote1 * cote2 ) /2.0)
        print (" Perimetre =", cote1 